botorch_community/models/np_regression.py

```diff
@@ -262,35 +262,38 @@ def __init__(
             forward pass.
         """
         super().__init__()
+        self.device = train_X.device
+
+        # self._validate_tensor_args(X=train_X, Y=train_Y)
         self.r_encoder = REncoder(
             x_dim + y_dim,
             r_dim,
             r_hidden_dims,
             activation=activation,
             init_func=init_func,
-        ).to(device)
+        ).to(self.device)
         self.z_encoder = ZEncoder(
             r_dim, z_dim, z_hidden_dims, activation=activation, init_func=init_func
-        ).to(device)
+        ).to(self.device)
         self.decoder = Decoder(
             x_dim + z_dim,
             y_dim,
             decoder_hidden_dims,
             activation=activation,
             init_func=init_func,
-        ).to(device)
-        self.train_X = train_X.to(device)
-        self.train_Y = train_Y.to(device)
+        ).to(self.device)
+        self.train_X = train_X.to(self.device)
+        self.train_Y = train_Y.to(self.device)
         self.n_context = n_context
         self.z_dim = z_dim
         self.z_mu_all = None
         self.z_logvar_all = None
         self.z_mu_context = None
         self.z_logvar_context = None
         if likelihood is None:
-            self.likelihood = GaussianLikelihood().to(device)
+            self.likelihood = GaussianLikelihood().to(self.device)
         else:
-            self.likelihood = likelihood.to(device)
+            self.likelihood = likelihood.to(self.device)

```

Comment:
Can't you just call self.to(device=train_X.device) once at the end of `__init__`?